In [1]:
import numpy as np
import pandas as pd
raw_data = {
"order_id": [
1001, 1002, 1003, 1004, 1005,
1006, 1007, 1008, 1009, 1010,
1011, 1012, 1013, 1014, 1015
],
"channel": [
"site", "app", "app", "marketplace", "site",
"app", "site", "marketplace", "app", "site",
"marketplace", "app", "site", "site", "marketplace"
],
"order_sum": [
1200, 2500, 1800, 3200, 900,
1500, 2800, 45000, 2100, 1700,
3600, 1100, 2400, 1900, 3300
],
"delivery_time": [
25, np.nan, 40, 55, 20,
30, np.nan, 70, 35, 45,
50, 18, 27, np.nan, 60
],
"rating": [
5, 4, 5, 3, 5,
4, 2, 1, 5, 4,
3, 5, 4, 5, 2
]
}
df = pd.DataFrame(raw_data) #преобразование словаря в DataFrame
print(df)
df.to_csv("online_store_orders.csv", index=False) #сохраняем файл в CSV

    order_id      channel  order_sum  delivery_time  rating
0       1001         site       1200           25.0       5
1       1002          app       2500            NaN       4
2       1003          app       1800           40.0       5
3       1004  marketplace       3200           55.0       3
4       1005         site        900           20.0       5
5       1006          app       1500           30.0       4
6       1007         site       2800            NaN       2
7       1008  marketplace      45000           70.0       1
8       1009          app       2100           35.0       5
9       1010         site       1700           45.0       4
10      1011  marketplace       3600           50.0       3
11      1012          app       1100           18.0       5
12      1013         site       2400           27.0       4
13      1014         site       1900            NaN       5
14      1015  marketplace       3300           60.0       2


## Задание 3.1 — формирование и сохранение данных

Сформирован исходный DataFrame `df` из словаря `raw_data`.
Таблица содержит 15 заказов и 5 столбцов: order_id, channel,
order_sum, delivery_time, rating. Данные сохранены в файл
online_store_orders.csv без индекса строк.

**Ответ на контрольный вопрос:**  
Параметр index=False используется, чтобы не записывать в CSV
служебный индекс строк (0, 1, 2, ...). Если его не указать,
в файле появится лишний безымянный столбец с номерами строк,
который при обратной загрузке превратится в обычный столбец
данных и создаст путаницу.

In [2]:
store_df = pd.read_csv("online_store_orders.csv") #загрузка CSV
print("Первые 5 строк:")
print(store_df.head())

print("\nПоследние 5 строк:")
print(store_df.tail()) #Последние 5 строк

print("\nРазмер таблицы:")
print(store_df.shape) #Размер таблицы

print("\nНазвания столбцов:")
print(store_df.columns) #Названия столбцов

print("\nИнформация о таблице:")
print(store_df.info()) #Информация о таблице

Первые 5 строк:
   order_id      channel  order_sum  delivery_time  rating
0      1001         site       1200           25.0       5
1      1002          app       2500            NaN       4
2      1003          app       1800           40.0       5
3      1004  marketplace       3200           55.0       3
4      1005         site        900           20.0       5

Последние 5 строк:
    order_id      channel  order_sum  delivery_time  rating
10      1011  marketplace       3600           50.0       3
11      1012          app       1100           18.0       5
12      1013         site       2400           27.0       4
13      1014         site       1900            NaN       5
14      1015  marketplace       3300           60.0       2

Размер таблицы:
(15, 5)

Названия столбцов:
Index(['order_id', 'channel', 'order_sum', 'delivery_time', 'rating'], dtype='object')

Информация о таблице:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 5 col

## Задание 4.1 — загрузка и первичный анализ

Файл online_store_orders.csv загружен в DataFrame `store_df`.
Первые и последние 5 строк выведены, размер таблицы — (15, 5).

**Ответы на контрольные вопросы:**
1. Количество строк: 15
2. Количество столбцов: 5
3. Числовые столбцы: order_id, order_sum, delivery_time, rating
4. Столбцы с пропущенными значениями: delivery_time
5. Количество пропущенных значений: 3

Столбец delivery_time имеет тип float64, потому что содержит
пропуски (NaN), а нецелые числа не могут храниться в int64.

In [3]:
selected_df = store_df[["order_sum", "delivery_time"]] #оставляем только 2 столбца
print(selected_df)

    order_sum  delivery_time
0        1200           25.0
1        2500            NaN
2        1800           40.0
3        3200           55.0
4         900           20.0
5        1500           30.0
6        2800            NaN
7       45000           70.0
8        2100           35.0
9        1700           45.0
10       3600           50.0
11       1100           18.0
12       2400           27.0
13       1900            NaN
14       3300           60.0


In [4]:
slice_df = store_df.iloc[3:8, [2, 3]] #строĸи с индеĸсами от 3 до 7 вĸлючительно и 2-3 столбцы
print(slice_df)

   order_sum  delivery_time
3       3200           55.0
4        900           20.0
5       1500           30.0
6       2800            NaN
7      45000           70.0


## Задания 5.1 и 5.2 — выборка данных

5.1: через store_df[["order_sum", "delivery_time"]] получена таблица
из двух столбцов.

5.2: через .iloc[3:8, [2, 3]] получены строки с индексами 3–7
и столбцы order_sum, delivery_time. Правая граница диапазона
в .iloc не включается, поэтому указано 8, а не 7.

In [5]:
premium_orders = store_df[(store_df["order_sum"] > 2000) & (store_df["rating"] == 5)] #заказы по условию
print(premium_orders)

   order_id channel  order_sum  delivery_time  rating
8      1009     app       2100           35.0       5


## Задание 5.3 — фильтрация по двум условиям

Через маску (store_df["order_sum"] > 2000) & (store_df["rating"] == 5)
найден один заказ, удовлетворяющий обоим условиям: order_id = 1009,
канал app, сумма 2100, время доставки 35 минут, рейтинг 5.

**Ответ на контрольный вопрос:**  
Используется &, а не and, потому что and работает с отдельными
значениями True/False и не умеет работать с целыми столбцами Pandas.
Оператор & выполняет поэлементное сравнение и возвращает массив
True/False, который можно применить к таблице.

In [6]:
print(store_df.isna().sum()) #Пропущенные значения

order_id         0
channel          0
order_sum        0
delivery_time    3
rating           0
dtype: int64


## Задание 6.1 — пропущенные значения

Через store_df.isna().sum() определено количество пропусков
в каждом столбце. Пропуски есть только в delivery_time — 3 значения.
Остальные столбцы заполнены полностью.

Ответ на вопрос: 3 заказа не имеют информации о времени доставки.

In [7]:
print(store_df.describe()) #Описательная статистиĸа

          order_id    order_sum  delivery_time     rating
count    15.000000     15.00000      12.000000  15.000000
mean   1008.000000   5000.00000      39.583333   3.800000
std       4.472136  11096.33144      16.654010   1.320173
min    1001.000000    900.00000      18.000000   1.000000
25%    1004.500000   1600.00000      26.500000   3.000000
50%    1008.000000   2100.00000      37.500000   4.000000
75%    1011.500000   3000.00000      51.250000   5.000000
max    1015.000000  45000.00000      70.000000   5.000000


## Задание 6.2 — описательная статистика

Через store_df.describe() получена сводка по числовым столбцам.

- order_sum: 15 наблюдений, mean = 5000, std = 11096, медиана = 2100,
  max = 45000.
- delivery_time: 12 наблюдений (3 пропуска не учитываются),
  mean = 39.58, медиана = 37.5, max = 70.
- rating: 15 наблюдений, mean = 3.8, диапазон 1–5.

Ответ на контрольный вопрос: столбец order_sum имеет max = 45000,
что сильно отличается от остальных значений (900–3600). Это выброс,
который сильно завышает среднее: среднее = 5000, а медиана = 2100.

In [8]:
mean_delivery = store_df["delivery_time"].mean() #среднее
median_delivery = store_df["delivery_time"].median() #медиана

print("Среднее:", mean_delivery)
print("Медиана:", median_delivery)

Среднее: 39.583333333333336
Медиана: 37.5


## Задание 6.3 — среднее и медиана времени доставки

Среднее время доставки = 39.58 минут, медиана = 37.5 минут.
Пропуски (NaN) автоматически игнорируются методами mean() и median().

Ответ на контрольный вопрос:  
Среднее и медиана отличаются, потому что считаются по-разному.
Среднее учитывает каждое значение и чувствительно к выбросам.
Медиана — это середина отсортированного ряда, она устойчива
к экстремальным значениям. В наших данных есть выброс 70 минут,
который тянет среднее вверх, поэтому среднее (39.58) немного
больше медианы (37.5).

In [9]:
print(store_df["rating"].value_counts()) #сколько раз встречается каждое уникальное значение

rating
5    6
4    4
3    2
2    2
1    1
Name: count, dtype: int64


## Задание 6.4 — распределение рейтингов

Через store_df["rating"].value_counts() получено распределение оценок:
- 5 → 6 заказов
- 4 → 4 заказа
- 3 → 2 заказа
- 2 → 2 заказа
- 1 → 1 заказ

Ответ на контрольный вопрос: наиболее часто встречается оценка 5 —
6 раз. Это самая высокая оценка, что говорит о преимущественно
положительных отзывах клиентов.

In [10]:
store_df["is_high_value"] = store_df["order_sum"] > 3000 #создание признака
print(store_df[["order_id", "order_sum", "is_high_value"]])

    order_id  order_sum  is_high_value
0       1001       1200          False
1       1002       2500          False
2       1003       1800          False
3       1004       3200           True
4       1005        900          False
5       1006       1500          False
6       1007       2800          False
7       1008      45000           True
8       1009       2100          False
9       1010       1700          False
10      1011       3600           True
11      1012       1100          False
12      1013       2400          False
13      1014       1900          False
14      1015       3300           True


## Задание 7.1 — производный признак is_high_value

Через store_df["is_high_value"] = store_df["order_sum"] > 3000
создан булев столбец. True — сумма больше 3000, False — меньше
или равна 3000.

Заказов с is_high_value = True: 4 (order_id 1004, 1008, 1011, 1015).

In [12]:
channel_stats = store_df.groupby("channel")["order_sum"].mean() #группировка по каналу продаж
print(channel_stats)

channel
app             1800.000000
marketplace    13775.000000
site            1816.666667
Name: order_sum, dtype: float64


## Задание 7.2 — группировка по каналу продаж

Через groupby("channel")["order_sum"].mean() рассчитана средняя
сумма заказа для каждого канала:
- app: 1800.00
- marketplace: 13775.00
- site: 1816.67

Средняя сумма marketplace в 7 раз выше остальных каналов.
Причина — заказ с суммой 45000, который попал в этот канал
и сильно завысил среднее. Без него средняя по marketplace
была бы около 3367.

In [13]:
channel_stats = channel_stats.sort_values(ascending=False) #сортировка результата в порядĸе убывания
print(channel_stats)

channel
marketplace    13775.000000
site            1816.666667
app             1800.000000
Name: order_sum, dtype: float64


## Задание 7.3 — сортировка

Через sort_values(ascending=False) результат отсортирован
по убыванию средней суммы заказа:
1. marketplace — 13775.00
2. site — 1816.67
3. app — 1800.00

## Задание 7.4 — интерпретация результатов

1. Средняя сумма заказа по каналам:
   - marketplace: 13775.00
   - site: 1816.67
   - app: 1800.00

2. Заказов с is_high_value = True: 4 (order_id 1004, 1008, 1011, 1015).

3. Пропущенных значений: 3 — все в delivery_time.

4. Влияние выброса: заказ с суммой 45000 (1008, marketplace)
   сильно завышает среднее по marketplace — 13775 против ~3367
   без него. Медиана по order_sum (2100) гораздо ниже среднего (5000)
   именно из-за этого выброса.

5. Различие среднего и медианного времени доставки:
   среднее 39.58, медиана 37.5. Разница небольшая — из-за одного
   выброса 70 минут.

**Итоговый вывод:**  
Средняя сумма заказа по каналам продаж сильно различается из-за
одного выброса: marketplace показывает 13775, тогда как site и app —
около 1800. При интерпретации средней суммы это нужно учитывать:
медиана (2100) более точно описывает типичный заказ, чем среднее (5000).
Среднее и медианное время доставки близки (39.58 и 37.5), что говорит
о стабильности этого показателя. 4 заказа из 15 имеют высокую стоимость
(> 3000), 3 заказа не содержат информации о времени доставки.

## Итоговый аналитический вывод

В наборе данных содержится 15 заказов интернет-магазина. В столбце
delivery_time обнаружено 3 пропущенных значения — по этим заказам
информация о времени доставки отсутствует. Среднее время доставки
составляет 39.58 минут, медианное — 37.5 минут; значения близки,
что говорит о стабильности показателя. Средняя сумма заказа по каналам
продаж составляет: marketplace — 13775 рублей, site — 1816.67 рублей,
app — 1800 рублей. При интерпретации средней суммы заказа необходимо
учитывать наличие выброса: один заказ на 45000 рублей сильно завышает
среднюю сумму как по marketplace, так и в целом по выборке — медиана
(2100) намного ниже среднего (5000). Всего 4 заказа из 15 имеют
высокую стоимость (более 3000 рублей).